### PR TITLE
Sk/prune sms

### PR DESCRIPTION
**Pull request by snopoke**
##### SUMMARY
Update SMS models to allow null values in foreign keys to events and create task to delete old event data.

https://dimagi-dev.atlassian.net/browse/IIO-429

##### RISK ASSESSMENT / QA PLAN
Apart from the DB migration this doesn't actually do anything. The task is not yet called from anywhere. On ICDS we will need to run it manually to delete old data before having it run automatically.

SQL query plans:

```sql
commcarehq=> explain UPDATE sms_sms sms SET messaging_subevent_id = NULL FROM sms_messagingsubevent as subevent LEFT JOIN sms_messagingevent as event ON event.id = subevent.parent_id WHERE sms.messaging_subevent_id = subevent.id and event.date between '2020-05-25' and '2020-06-25';
                                                                               QUERY PLAN                                                                                
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Update on sms_sms sms  (cost=24493333.51..30461149.31 rows=11323274 width=3195)
   ->  Merge Join  (cost=24493333.51..30461149.31 rows=11323274 width=3195)
         Merge Cond: (sms.messaging_subevent_id = subevent.id)
         ->  Index Scan using sms_sms_af811fea on sms_sms sms  (cost=0.56..5540146.85 rows=45417656 width=3183)
         ->  Materialize  (cost=24493269.98..24750903.71 rows=51526747 width=16)
               ->  Sort  (cost=24493269.98..24622086.84 rows=51526747 width=16)
                     Sort Key: subevent.id
                     ->  Merge Join  (cost=2015.14..17377226.36 rows=51526747 width=16)
                           Merge Cond: (subevent.parent_id = event.id)
                           ->  Index Scan using sms_messagingsubevent_6be37982 on sms_messagingsubevent subevent  (cost=0.57..8175274.69 rows=206673808 width=14)
                           ->  Index Scan using sms_messagingevent_pkey on sms_messagingevent event  (cost=0.57..8107771.03 rows=52482619 width=10)
                                 Filter: ((date >= '2020-05-25 00:00:00+00'::timestamp with time zone) AND (date <= '2020-06-25 00:00:00+00'::timestamp with time zone))
(12 rows)

commcarehq=> 
commcarehq=> explain DELETE FROM sms_messagingsubevent as subevent USING sms_messagingevent as event WHERE subevent.parent_id = event.id AND event.date between '2020-05-25' and '2020-06-25';
                                                                      QUERY PLAN                                                                       
-------------------------------------------------------------------------------------------------------------------------------------------------------
 Delete on sms_messagingsubevent subevent  (cost=1957.26..17017760.94 rows=51526747 width=12)
   ->  Merge Join  (cost=1957.26..17017760.94 rows=51526747 width=12)
         Merge Cond: (subevent.parent_id = event.id)
         ->  Index Scan using sms_messagingsubevent_6be37982 on sms_messagingsubevent subevent  (cost=0.57..7925472.22 rows=206673808 width=10)
         ->  Index Scan using sms_messagingevent_pkey on sms_messagingevent event  (cost=0.57..7997182.22 rows=52482619 width=10)
               Filter: ((date >= '2020-05-25 00:00:00+00'::timestamp with time zone) AND (date <= '2020-06-25 00:00:00+00'::timestamp with time zone))
(6 rows)

commcarehq=> 
commcarehq=> explain DELETE FROM sms_messagingevent WHERE date between '2020-05-25' and '2020-06-25';
                                                                     QUERY PLAN                                                                      
-----------------------------------------------------------------------------------------------------------------------------------------------------
 Delete on sms_messagingevent  (cost=0.57..1995630.26 rows=52482619 width=6)
   ->  Index Scan using sms_messagingevent_5fc73231 on sms_messagingevent  (cost=0.57..1995630.26 rows=52482619 width=6)
         Index Cond: ((date >= '2020-05-25 00:00:00+00'::timestamp with time zone) AND (date <= '2020-06-25 00:00:00+00'::timestamp with time zone))
(3 rows)
```

### CODE REVIEW

File: corehq/apps/sms/migrations/0043_auto_20200622_1243.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        migrations.AlterField(
+            model_name='messagingevent',
+            name='status',
+            field=models.CharField(choices=[('PRG', 'In Progress'), ('CMP', 'Completed'), ('NOT', 'Not Completed'), ('ERR', 'Error'), ('SND', 'Email Sent'), ('DEL', 'Email Delivered')], max_length=3),
+        ),
+        migrations.AlterField(
+            model_name='messagingsubevent',
+            name='status',
+            field=models.CharField(choices=[('PRG', 'In Progress'), ('CMP', 'Completed'), ('NOT', 'Not Completed'), ('ERR', 'Error'), ('SND', 'Email Sent'), ('DEL', 'Email Delivered')], max_length=3),
+        ),
```
Comment (snopoke) [inline corehq/apps/sms/migrations/0043_auto_20200622_1243.py:45]: These are migrations unrelated to this code and are noops

File: corehq/apps/sms/migrations/0042_auto_20200622_1243.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('sms', '0041_email'),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='InfobipBackend',
```
Comment (mkangia) [inline corehq/apps/sms/migrations/0042_auto_20200622_1243.py]: these were already added in https://github.com/dimagi/commcare-hq/pull/27851/files so you might need to update this here.

**Review by mkangia — COMMENTED** [review]
Thanks for adding this @snopoke

**Review by millerdev — APPROVED** [review]

File: custom/icds/tasks/sms.py [hunk 3]
```diff
@@ -118,3 +124,35 @@ def send_sms_limit_exceeded_alert(sender, instance, **kwargs):
             """).format(env=settings.SERVER_ENVIRONMENT, date=instance.date, sms_limit=sms_daily_limit)
         send_html_email_async.delay(subject, recipients, message,
                 email_from=settings.DEFAULT_FROM_EMAIL)
+
+
+@task
+def delete_sms_events(start_date, end_date):
```
Comment (sravfeyn) [inline custom/icds/tasks/sms.py:130]: Should we add a validation or a warning on `end_date` to not delete very recent ones?

**Review by sravfeyn — APPROVED** [review]

**Pull request merged**